File: frb_example/pure_dart/lib/src/rust/api/dart_opaque.dart
// ignore_for_file: invalid_use_of_internal_member, unused_import

import '../frb_generated.dart';
import 'package:collection/collection.dart';
import 'package:flutter_rust_bridge/flutter_rust_bridge_for_generated.dart';
import 'package:freezed_annotation/freezed_annotation.dart' hide protected;
part 'dart_opaque.freezed.dart';

Future<String> asyncAcceptDartOpaque({required Object opaque, dynamic hint}) =>
    RustLib.instance.api.asyncAcceptDartOpaque(opaque: opaque, hint: hint);

Future<Object> loopBack({required Object opaque, dynamic hint}) =>
    RustLib.instance.api.loopBack(opaque: opaque, hint: hint);

Future<Object?> loopBackOption({required Object opaque, dynamic hint}) =>
    RustLib.instance.api.loopBackOption(opaque: opaque, hint: hint);

Future<ObjectArray1> loopBackArray({required Object opaque, dynamic hint}) =>
    RustLib.instance.api.loopBackArray(opaque: opaque, hint: hint);

Future<List<Object>> loopBackVec({required Object opaque, dynamic hint}) =>
    RustLib.instance.api.loopBackVec(opaque: opaque, hint: hint);

Future<void> loopBackOptionGet({Object? opaque, dynamic hint}) =>
    RustLib.instance.api.loopBackOptionGet(opaque: opaque, hint: hint);

Future<void> loopBackArrayGet({required ObjectArray1 opaque, dynamic hint}) =>
    RustLib.instance.api.loopBackArrayGet(opaque: opaque, hint: hint);

Future<void> loopBackVecGet({required List<Object> opaque, dynamic hint}) =>
    RustLib.instance.api.loopBackVecGet(opaque: opaque, hint: hint);

/// [DartWrapObject] cannot be obtained
/// on a thread other than the thread it was created on.
Future<void> panicUnwrapDartOpaque({required Object opaque, dynamic hint}) =>
    RustLib.instance.api.panicUnwrapDartOpaque(opaque: opaque, hint: hint);

Future<DartOpaqueNested> createNestedDartOpaque(
        {required Object opaque1, required Object opaque2, dynamic hint}) =>
    RustLib.instance.api
        .createNestedDartOpaque(opaque1: opaque1, opaque2: opaque2, hint: hint);

Future<void> getNestedDartOpaque(
        {required DartOpaqueNested opaque, dynamic hint}) =>
    RustLib.instance.api.getNestedDartOpaque(opaque: opaque, hint: hint);

Future<EnumDartOpaque> createEnumDartOpaque(
        {required Object opaque, dynamic hint}) =>
    RustLib.instance.api.createEnumDartOpaque(opaque: opaque, hint: hint);

Future<void> getEnumDartOpaque(
        {required EnumDartOpaque opaque, dynamic hint}) =>
    RustLib.instance.api.getEnumDartOpaque(opaque: opaque, hint: hint);

Future<void> setStaticDartOpaque({required Object opaque, dynamic hint}) =>
    RustLib.instance.api.setStaticDartOpaque(opaque: opaque, hint: hint);

Future<void> dropStaticDartOpaque({dynamic hint}) =>
    RustLib.instance.api.dropStaticDartOpaque(hint: hint);

class ObjectArray1 extends NonGrowableListView<Object> {
  static const arraySize = 1;
  ObjectArray1(List<Object> inner)
      : assert(inner.length == arraySize),
        super(inner);
  ObjectArray1.unchecked(List<Object> inner) : super(inner);
  ObjectArray1.init(Object fill) : super(List<Object>.filled(arraySize, fill));
}

class DartOpaqueNested {
  final Object first;
  final Object second;

  const DartOpaqueNested({
    required this.first,
    required this.second,
  });

  @override
  int get hashCode => first.hashCode ^ second.hashCode;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is DartOpaqueNested &&
          runtimeType == other.runtimeType &&
          first == other.first &&
          second == other.second;
}

@freezed
sealed class EnumDartOpaque with _$EnumDartOpaque {
  const factory EnumDartOpaque.primitive(
    int field0,
  ) = EnumDartOpaque_Primitive;
  const factory EnumDartOpaque.opaque(
    Object field0,
  ) = EnumDartOpaque_Opaque;
}

File: frb_example/pure_dart/rust/src/api/dart_opaque.rs
use anyhow::Result;
use flutter_rust_bridge::support::lazy_static;
use flutter_rust_bridge::{frb, DartOpaque};

// TODO about sync
// pub fn sync_accept_dart_opaque(opaque: DartOpaque) -> SyncReturn<String> {
//     drop(opaque);
//     SyncReturn("test".to_owned())
// }

pub fn async_accept_dart_opaque(opaque: DartOpaque) -> String {
    drop(opaque);
    "async test".to_owned()
}

pub fn loop_back(opaque: DartOpaque) -> DartOpaque {
    opaque
}

pub fn loop_back_option(opaque: DartOpaque) -> Option<DartOpaque> {
    Some(opaque)
}

pub fn loop_back_array(opaque: DartOpaque) -> [DartOpaque; 1] {
    [opaque]
}

pub fn loop_back_vec(opaque: DartOpaque) -> Vec<DartOpaque> {
    vec![opaque]
}

pub fn loop_back_option_get(opaque: Option<DartOpaque>) {}

pub fn loop_back_array_get(opaque: [DartOpaque; 1]) {}

pub fn loop_back_vec_get(opaque: Vec<DartOpaque>) {}

/// [DartWrapObject] cannot be obtained
/// on a thread other than the thread it was created on.
pub fn panic_unwrap_dart_opaque(opaque: DartOpaque) {
    let handle = opaque.try_unwrap().unwrap();
}

pub enum EnumDartOpaque {
    Primitive(i32),
    Opaque(DartOpaque),
}

pub struct DartOpaqueNested {
    pub first: DartOpaque,
    pub second: DartOpaque,
}

// TODO about sync
// pub fn sync_loopback(opaque: DartOpaque) -> SyncReturn<DartOpaque> {
//     SyncReturn(opaque)
// }
//
// pub fn sync_option_loopback(opaque: Option<DartOpaque>) -> SyncReturn<Option<DartOpaque>> {
//     SyncReturn(opaque)
// }
//
// pub fn sync_option_dart_opaque(opaque: DartOpaque) -> Result<SyncReturn<Option<DartOpaque>>> {
//     Ok(SyncReturn(Some(opaque)))
// }

pub fn create_nested_dart_opaque(opaque1: DartOpaque, opaque2: DartOpaque) -> DartOpaqueNested {
    DartOpaqueNested {
        first: opaque1,
        second: opaque2,
    }
}

pub fn get_nested_dart_opaque(opaque: DartOpaqueNested) {}

pub fn create_enum_dart_opaque(opaque: DartOpaque) -> EnumDartOpaque {
    EnumDartOpaque::Opaque(opaque)
}

pub fn get_enum_dart_opaque(opaque: EnumDartOpaque) {}

lazy_static! {
    static ref DART_OPAQUE: Mutex<Option<DartOpaque>> = Default::default();
}

pub fn set_static_dart_opaque(opaque: DartOpaque) {
    *DART_OPAQUE.lock().unwrap() = Some(opaque);
}

pub fn drop_static_dart_opaque() {
    drop(DART_OPAQUE.lock().unwrap().take());
}
